From e991b6c7d9c31187e3b9830759b4e26fa10b7e3c Mon Sep 17 00:00:00 2001
From: cyberCuber07 <n.eigocuber@gmail.com>
Date: Sat, 25 Oct 2025 11:00:47 +0200
Subject: [fix] Make compile on unix systems

Following fixing issues found in JFM code -
 - this patch makes the code compliant with cpp
standard as msvc treats it loosely.
Those issues are:
  - making variables with names same as template arguments
  - remove 'typename' keyword in alias' definees

Signed-off-by: cyberCuber07 <n.eigocuber@gmail.com>
---
 NumericStorm/headers/Fitting/ModelBase.hpp                | 8 ++++----
 .../BasicSimplex/BasicSimplexOptimizerSettings.hpp        | 6 ++++--
 NumericStorm/headers/Utils/type_list.hpp                  | 7 ++++++-
 3 files changed, 14 insertions(+), 7 deletions(-)

diff --git a/NumericStorm/headers/Fitting/ModelBase.hpp b/NumericStorm/headers/Fitting/ModelBase.hpp
index e7803b5..a57f01a 100644
--- a/NumericStorm/headers/Fitting/ModelBase.hpp
+++ b/NumericStorm/headers/Fitting/ModelBase.hpp
@@ -11,13 +11,13 @@ namespace NumericStorm
 {
 namespace Fitting
 {
-template <size_t parameter_size, class AuxilaryParameters = AdditionalParameters>
+template <size_t parameter_size_t, class AuxilaryParameters = AdditionalParameters>
 class ModelBase
 {
 public:
-    static const size_t parameter_size = parameter_size;
+    static const size_t parameter_size = parameter_size_t;
     using AuxParameters = AuxilaryParameters;
-    using ModelFunction = std::function<void(Data&, const Parameters<parameter_size>&, const AuxParameters&)>;
+    using ModelFunction = std::function<void(Data&, const Parameters<parameter_size_t>&, const AuxParameters&)>;
     
     ModelBase() = default;
 
@@ -27,7 +27,7 @@ public:
    
     virtual ~ModelBase() = default;
 
-    void operator()(Data& arguments, const Parameters<parameter_size>& parameters, const AuxParameters& additionalParameters) const {
+    void operator()(Data& arguments, const Parameters<parameter_size_t>& parameters, const AuxParameters& additionalParameters) const {
         m_model(arguments, parameters, additionalParameters);
     }
 
diff --git a/NumericStorm/headers/Fitting/Nealder-Mead/BasicSimplex/BasicSimplexOptimizerSettings.hpp b/NumericStorm/headers/Fitting/Nealder-Mead/BasicSimplex/BasicSimplexOptimizerSettings.hpp
index a5b97b0..b6362f2 100644
--- a/NumericStorm/headers/Fitting/Nealder-Mead/BasicSimplex/BasicSimplexOptimizerSettings.hpp
+++ b/NumericStorm/headers/Fitting/Nealder-Mead/BasicSimplex/BasicSimplexOptimizerSettings.hpp
@@ -57,11 +57,13 @@ namespace NumericStorm::Fitting
 		friend class BasicSimplexOptimizerSettingsBuilderBase;
 		friend class BasicSimplexSettingsBuilder;
 	
-	protected:
-		
 	
 		template<class BuildingType, OptimizerSettings Settings>
+#ifdef __clang__
+		class BasicSimplexOptimizerSettingsBuilderBase : public SimplexOptimizerSettings<M>::template SimplexOptimizerSettingsBuilderBase<BuildingType, Settings> {
+#else
 		class BasicSimplexOptimizerSettingsBuilderBase : public SimplexOptimizerSettings<M>::SimplexOptimizerSettingsBuilderBase<BuildingType, Settings> {
+#endif
 		public:
 
 			BuildingType& addOperationSettings(const SettingsPair& settings)
diff --git a/NumericStorm/headers/Utils/type_list.hpp b/NumericStorm/headers/Utils/type_list.hpp
index 83d82a3..53d92f6 100644
--- a/NumericStorm/headers/Utils/type_list.hpp
+++ b/NumericStorm/headers/Utils/type_list.hpp
@@ -16,8 +16,13 @@ namespace NumericStorm::Utils
 	template<class Head, class... Tail>
 	struct pop_front<type_list<Head, Tail...>>
 	{
+#if !defined(JFM_PLATFORM_NIX)
 		using out = typename Head;
 		using type = typename type_list<Tail...>;
+#else
+		using out = Head;
+		using type = type_list<Tail...>;
+#endif
 	};
 
 	template<>
@@ -33,4 +38,4 @@ namespace NumericStorm::Utils
 	template<class List>
 	using pop_front_o = typename pop_front<List>::out;
 
-}
\ No newline at end of file
+}
-- 
2.43.0

